src/group_theory/subsemigroup/membership.lean

```diff
@@ -0,0 +1,131 @@
+/-
+Copyright (c) 2022 Jireh Loreaux. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+import group_theory.subsemigroup.operations
+
+/-!
+# Subsemigroups: membership criteria
+
+In this file we prove various facts about membership in a subsemigroup.
+The intent it to mimic `group_theory/submonoid/membership`, but currently this file is mostly a
+stub and only provides rudimentary support.
+
+* `mem_supr_of_directed`, `coe_supr_of_directed`, `mem_Sup_of_directed_on`,
+  `coe_Sup_of_directed_on`: the supremum of a directed collection of subsemigroup is their union.
+
+## TODO
+
+* Define the `free_semigroup` generated by a set. This might require some rather substantial
+  additions to low-level API. For example, developing the subtype of nonempty lists, then defining
+  a product on nonempty lists, powers where the exponent is a positive natural, et cetera.
+  Another option would be to define the `free_semigroup` as the subsemigroup (pushed to be a
+  semigroup) of the `free_monoid` consisting of non-identity elements.
+
+## Tags
+subsemigroup
+-/
+
+variables {M A B : Type*}
+
+/-
+Here, we can't have various product lemmas because we don't have a `1 : M`. Perhaps we need to
+define a type of nonempty lists, and then products on those?
+-/
+
+section non_assoc
+variables [has_mul M]
+
+open set
+
+namespace subsemigroup
+
+-- TODO: this section can be generalized to `[mul_mem_class B M] [complete_lattice B]`
+-- such that `complete_lattice.le` coincides with `set_like.le`
```

Comment:
Is there any reason not to do this now?

Reply:
Yes, this is actually a nontrivial change. In particular, we would need to create a new type class `complete_set_like` which extends both `set_like` and `complete_lattice` in a suitable way (since both of these already define a `has_le`). That work is outside the scope of this PR.